Add Python client

isinyaaa opened the pull request:
Add a preliminary Python client implementation.

## Description
I've based my implementation on #68 and our OpenAPI definition.

## How Has This Been Tested?
There are simple tests included. I still have to add instructions to run, a CI job to run them
automatically, and docstrings.

To test, first make sure you have at least Python 3.10 installed, if you don't:
1. Setup [`pyenv`](https://github.com/pyenv/pyenv)
2. `pyenv install 3.10.13`

Now, install [`nox`](https://github.com/wntrblm/nox), then, to test, follow the instructions on the README.md file.

## TODO:
- [x] Document `nox`, etc on the README
- [x] Build Docs
- [x] Add CI jobs for building docs and running tests
- [ ] Release the first version

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits and have meaningful messages
- [ ] The author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

I don't know if it makes sense to squash this, we can discuss that.

dhirajsb commented:
@isinyaaa I think one of the suggestions from @rimolive from DSP team was that we could re-use mlmd base classes and extend our types from them. 
Would that make it possible to inherit common properties and perhaps make it easier to implement the API somehow? 
I am not sure about all the implementation details. @rimolive can you provide more information about what you were thinking?

isinyaaa commented:
v2 updates:
- Moved the code into a `src` folder as per guidelines.
- Removed `map` submodule: let's keep things simple with a single interface and helper methods.
- Moved model registry client into submodule
- Created a new interface for prefixable types
- Simplified artifact state mapping (removed bidict dependency)
- Added simple docstrings
- Created a noxfile to simplify running tasks
- Better separation of core functionality from MLMD integration/dependency
- Dealt with a few TODOs
- Added pytest coverage plugin
- Improved test coverage (now at 87%) and logic

dhirajsb reviewed (dismissed):
lgtm

Just to clarify, is the primary purpose of the store wrapper type for parity with the Golang client's service layer API?

isinyaaa commented:
> lgtm
> 
> Just to clarify, is the primary purpose of the store wrapper type for parity with the Golang client's service layer API?

I'm focusing on parity on the base type definitions and on the registry client. The store wrapper should simply provide a useful interface to gRPC focusing on our use case (upserting and querying single protos).

isinyaaa commented on clients/python/src/model_registry/types/base.py:
```diff
@@ -0,0 +1,145 @@
+    id: str
+    name: str
+    external_id: Optional[str]
+    create_time_since_epoch: int
+    last_update_time_since_epoch: int
```
TBH I can't really see the usefulness of providing these fields to the client at the cost of an additional query. The only potential use case I see is some way of sorting by timestamp, but `ListOptions` already provides this functionality when querying in bulk on the server side.

dhirajsb replied:
Yeah, unfortunately gRPC only returns the ids on create calls. Let's go ahead and do the extra call because wewill have default values for some _system_ properties beside these timestamps. And, the client library should return those defaults to the user. 
For example, we'll soon need to add a creator/owner username property to show who created which models, etc. in the UI. That property will also be needed for RBAC in the future.

isinyaaa commented:
Changes in v3:
- Added sphinx to build docs
- Added basic usage on README
- Added attrs
- Made exceptions a common submodule
- Brought back client as single file module as we don't need much more right now
- Fixed license string
- Added CI to build docs and test
- Added release commit
- Fixed (most) `get_*` calls to automatically populate references to other proto objects. (Artifacts are still pending)
- Fixed ListOptions
- Fixed `ModelRegistry` constructor to take in SSL connection args
- Renamed `*Error` -> `*Exception` following python conventions.
- Some more improvements to tests

dtrifiro commented on clients/python/pyproject.toml:
```diff
@@ -0,0 +1,44 @@
+[tool.poetry]
```
Did you encounter any issues so far by using poetry?

isinyaaa replied:
I had some issues with Python versions, and it's also not as clear how to get proper integration with an editor (I've been using `VenvSelector` on vim with some hiccups on my Mac).

dtrifiro commented on clients/python/src/model_registry/__init__.py:
```diff
@@ -0,0 +1,12 @@
+"""Main package for the ODH model registry."""
+
+__version__ = "0.1.0"
```
Also related to the above: I find it rather cumbersome to have version numbers hardcoded in code and/or `pyproject.toml`.

To solve this issue, I like building with setuptools and using [`setuptools_scm`](https://github.com/pypa/setuptools_scm/), which makes it easy to retrieve version information from git. 

If you like to stick with poetry, something that **might** work with poetry is this [`poetry-dynamic-versioning`](https://github.com/mtkennerly/poetry-dynamic-versioning/), although I have not used it before. If you want to do some reading, there's quite a few issues in the poetry repo related to this problem.

isinyaaa replied:
I'd had a look at that and thought it was unnecessary for now, as we are still iterating on an alpha version. This plugin you sent seems to rely on git tags, which we can't use, unless we share tags for the go MR as well as the python client.

dtrifiro commented on clients/python/src/model_registry/client.py:
```diff
@@ -0,0 +1,408 @@
+        Returns:
+            Sequence[RegisteredModel]: Registered models.
+        """
+        if options is None:
+            mlmd_options = None
+        else:
+            mlmd_options = options.as_mlmd_list_options()
```
```suggestion
       mlmd_options = options.as_mlmd_list_options() if options else None
```

dtrifiro commented on clients/python/src/model_registry/client.py:
```diff
@@ -0,0 +1,408 @@
+        model_version_id: Optional[str] = None,
+        options: Optional[ListOptions] = None,
+    ) -> Sequence[ModelArtifact]:
+        if options is None:
+            mlmd_options = MLMDListOptions()
+        else:
+            mlmd_options = options.as_mlmd_list_options()
```
Same as above

dtrifiro commented on clients/python/src/model_registry/exceptions.py:
```diff
@@ -0,0 +1,32 @@
+class StoreException(Exception):
+    """Storage related error."""
+
+    pass
```
This pass (and all of the following) are not required

dtrifiro commented on clients/python/src/model_registry/types/base.py:
```diff
@@ -0,0 +1,204 @@
+        for key, value in py_props.items():
+            if value is None:
+                continue
+            if isinstance(value, bool):
+                mlmd_props[key].bool_value = value
+            elif isinstance(value, int):
+                mlmd_props[key].int_value = value
+            elif isinstance(value, float):
+                mlmd_props[key].double_value = value
+            elif isinstance(value, str):
+                mlmd_props[key].string_value = value
+            else:
+                raise Exception(f"Unsupported type: {type(value)}")
```
Seems like a good place to use structural pattern matching: https://docs.python.org/3/whatsnew/3.10.html?highlight=match#pep-634-structural-pattern-matching

isinyaaa replied:
I had tried this before only to get `Irrefutable pattern is allowed only for the last case statement`, but now I see [I should be matching against an instance](https://stackoverflow.com/questions/72295812/python-match-case-by-type-of-value) and not the type itself.

dtrifiro commented on clients/python/src/model_registry/types/contexts.py:
```diff
@@ -0,0 +1,107 @@
+class BaseContext(ProtoBase, ABC):
+    """Abstract base class for all contexts."""
+
+    pass
```
useless pass statement

dtrifiro commented on clients/python/tests/store/test_wrapper.py:
```diff
@@ -0,0 +1,131 @@
+
+
+def create_artifact(store_wrapper: MLMDStore, name: str) -> Artifact:
+    """Create an artifact with the given name.
+
+    Args:
+        store_wrapper (MLMDStore): MLMD store wrapper.
+        name (str): Name of the artifact.
+
+    Returns:
+        Artifact: Created artifact.
+    """
+    art_type = ArtifactType()
+    art_type.name = name
+
+    art = Artifact()
+    art.name = name
+    art.type_id = store_wrapper._mlmd_store.put_artifact_type(art_type)
+
+    return art
+
+
+def create_context(store_wrapper: MLMDStore, name: str) -> Context:
+    """Create a context with the given name.
+
+    Args:
+        store_wrapper (MLMDStore): MLMD store wrapper.
+        name (str): Name of the context type.
+
+    Returns:
+        Context: Created context.
+    """
+    ctx_type = ContextType()
+    ctx_type.name = name
+
+    ctx = Context()
+    ctx.name = name
+    ctx.type_id = store_wrapper._mlmd_store.put_context_type(ctx_type)
+
+    return ctx
```
These could be pytest fixtures

isinyaaa replied:
I couldn't find a simple way to parameterize names for those, but now I see this is not necessary.

dtrifiro reviewed (approved):
Looks great!

tarilabs commented on clients/python/pyproject.toml:
```diff
@@ -0,0 +1,44 @@
+[tool.poetry]
+name = "model-registry"
+version = "0.1.0"
+description = "Client for Red hat OpenDataHub Model Registry"
```
```suggestion
description = "Client for OpenDataHub Model Registry"
```

Very minor, but I think being the community version, they would prefer not to prefix it, to inadvertently give wrong impression?
If you want to keep it, it's fine too, but capital H per brand guidelines :) `Red Hat`

isinyaaa replied:
Oh MB :) thx for the suggestion.

tarilabs reviewed (dismissed):
Imho you can consider Dhiraj and Daniele's comments in further PR, if you prefer to merge this now, and could be further improved directly on main for their suggestions and further dev work.

For the records for QE and others;
I've tested this both in DevContainer (I personally find that convenient for IDE support):

![Screenshot 2023-11-06 at 14 59 20](https://github.com/opendatahub-io/model-registry/assets/1699252/122f248c-6b3a-4ddb-bd30-1d8975a43732)

and Rosetta emulation:
```
$env /usr/bin/arch -x86_64 /bin/zsh --login
# then
poetry install
poetry build
poetry run pytest
```

isinyaaa commented on clients/python/src/model_registry/store/wrapper.py:
```diff
@@ -0,0 +1,314 @@
+        if type_name in self._type_ids:
+            return self._type_ids[type_name]
+
+        if isinstance(mlmd_pt, Artifact):
+            try:
+                _type = self._mlmd_store.get_artifact_type(type_name)
+            except errors.NotFoundError as e:
+                raise TypeNotFoundException(
+                    f"Artifact type {type_name} does not exist"
+                ) from e
+            except errors.InternalError as e:
+                raise ServerException(
+                    f"Couldn't get artifact type {type_name} from MLMD store"
+                ) from e
+        elif isinstance(mlmd_pt, Context):
+            try:
+                _type = self._mlmd_store.get_context_type(type_name)
+            except errors.NotFoundError as e:
+                raise TypeNotFoundException(
+                    f"Context type {type_name} does not exist"
+                ) from e
+            except errors.InternalError as e:
+                raise ServerException(
+                    f"Couldn't get context type {type_name} from MLMD store"
+                ) from e
+        else:
+            raise UnsupportedTypeException(f"Unsupported type: {type_name}")
```
Thx for the suggestion, I didn't bother too much with that as we're only dealing with those two. There must also be better ways of mapping/unmapping.

isinyaaa commented:
Changes in v4:
- Applied suggestions
- Used structural pattern matching where possible

isinyaaa commented:
Changes in v5:
- Dropped structural pattern matching to keep compatibility with 3.9.

tarilabs commented:
closing this one to trigger a rebase 👀
